### PR TITLE
[Mailer] Add Infobip bridge

### PR DESCRIPTION
**Pull request by B-Galati**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | symfony/symfony-docs#16760

Hello,

Add Infobip as new email provider for Symfony Mailer!

Co-authored with @jbdelhommeau :+1:

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -2436,6 +2436,7 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
             SendinblueTransportFactory::class => 'mailer.transport_factory.sendinblue',
             SesTransportFactory::class => 'mailer.transport_factory.amazon',
             OhMySmtpTransportFactory::class => 'mailer.transport_factory.ohmysmtp',
+            InfobipTransportFactory::class => 'mailer.transport_factory.infobip',
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: can you please sort them alphabetically? And please move OhMySmtpTransportFactory, too, thanks

File: src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+6.1.0
+-----
+
+ * Added the bridge
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md]:
```suggestion
6.1
---

 * Add the bridge
```

File: src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipApiTransport.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+use Symfony\Component\Mime\Part\Multipart\FormDataPart;
+use Symfony\Contracts\HttpClient\Exception\DecodingExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @see https://www.infobip.com/docs/api#channels/email/send-email
+ */
+class InfobipApiTransport extends AbstractApiTransport
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipApiTransport.php]:
```suggestion
final class InfobipApiTransport extends AbstractApiTransport
```

File: src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipSmtpTransport.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Infobip\Transport;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
+
+class InfobipSmtpTransport extends EsmtpTransport
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipSmtpTransport.php]:
```suggestion
final class InfobipSmtpTransport extends EsmtpTransport
```

File: src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipTransportFactory.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Symfony\Component\Mailer\Bridge\Infobip\Transport;
+
+use Symfony\Component\Mailer\Exception\IncompleteDsnException;
+use Symfony\Component\Mailer\Exception\UnsupportedSchemeException;
+use Symfony\Component\Mailer\Transport\AbstractTransportFactory;
+use Symfony\Component\Mailer\Transport\Dsn;
+use Symfony\Component\Mailer\Transport\TransportInterface;
+
+class InfobipTransportFactory extends AbstractTransportFactory
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Infobip/Transport/InfobipTransportFactory.php]:
```suggestion
final class InfobipTransportFactory extends AbstractTransportFactory
```

File: src/Symfony/Component/Mailer/Exception/UnsupportedSchemeException.php [hunk 6]
```diff
@@ -56,6 +56,10 @@ class UnsupportedSchemeException extends LogicException
             'class' => Bridge\OhMySmtp\Transport\OhMySmtpTransportFactory::class,
             'package' => 'symfony/oh-my-smtp-mailer',
         ],
+        'infobip' => [
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Exception/UnsupportedSchemeException.php]: Can you please sort this alphabetically? Thanks

Please move ohmysmtp, too

Thanks

File: src/Symfony/Component/Mailer/Tests/Exception/UnsupportedSchemeExceptionTest.php [hunk 7]
```diff
@@ -43,6 +44,7 @@ public static function setUpBeforeClass(): void
             SendinblueTransportFactory::class => false,
             OhMySmtpTransportFactory::class => false,
             SesTransportFactory::class => false,
+            InfobipTransportFactory::class => false,
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Tests/Exception/UnsupportedSchemeExceptionTest.php]: Please sort, thanks

File: src/Symfony/Component/Mailer/Tests/Exception/UnsupportedSchemeExceptionTest.php [hunk 8]
```diff
@@ -70,6 +72,7 @@ public function messageWhereSchemeIsPartOfSchemeToPackageMapProvider(): \Generat
         yield ['sendinblue', 'symfony/sendinblue-mailer'];
         yield ['ohmysmtp', 'symfony/oh-my-smtp-mailer'];
         yield ['ses', 'symfony/amazon-mailer'];
+        yield ['infobip', 'symfony/infobip-mailer'];
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Tests/Exception/UnsupportedSchemeExceptionTest.php]: Please sort, thanks

File: src/Symfony/Component/Mailer/Transport.php [hunk 9]
```diff
@@ -52,6 +53,7 @@ final class Transport
         SendinblueTransportFactory::class,
         OhMySmtpTransportFactory::class,
         SesTransportFactory::class,
+        InfobipTransportFactory::class,
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Transport.php]: Please sort, thanks

**Review by OskarStark — COMMENTED** [review]
Some minor comments, thank you

**Comment by B-Galati** [comment]
@OskarStark Thanks! All fixed, hope I did not forget anything :+1:

**Comment by B-Galati** [comment]
just done a little update to fix the SMTP client, all good now.

**Comment by B-Galati** [comment]
Rebased with last commit to check if it fixes psalm error.

**Comment by B-Galati** [comment]
@OskarStark I know it's a bit late but do you know if it could be released with 6.1?

**Comment by fabpot** [comment]
We’ve been in feature freeze for more than a month now, so it’ll be for 6.2 unfortunately.

**Comment by B-Galati** [comment]
Thank you, I will change the branch target then :+1:

File: src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+6.1
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md]: 6.2

File: src/Symfony/Component/Mailer/Bridge/Infobip/Tests/Transport/InfobipApiTransportFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Infobip/Tests/Transport/InfobipApiTransportFactoryTest.php]: all such lines should be removed

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 12]
```diff
@@ -2428,14 +2428,15 @@ private function registerMailerConfiguration(array $config, ContainerBuilder $co
...
             MailgunTransportFactory::class => 'mailer.transport_factory.mailgun',
             MailjetTransportFactory::class => 'mailer.transport_factory.mailjet',
             MandrillTransportFactory::class => 'mailer.transport_factory.mailchimp',
+            OhMySmtpTransportFactory::class => 'mailer.transport_factory.ohmysmtp',
             PostmarkTransportFactory::class => 'mailer.transport_factory.postmark',
             SendgridTransportFactory::class => 'mailer.transport_factory.sendgrid',
             SendinblueTransportFactory::class => 'mailer.transport_factory.sendinblue',
             SesTransportFactory::class => 'mailer.transport_factory.amazon',
-            OhMySmtpTransportFactory::class => 'mailer.transport_factory.ohmysmtp',
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Rebasing on current 6.2 should cleanup the diff here

File: src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md (hunk 10 above)
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Infobip/CHANGELOG.md]: 6.2

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
@B-Galati Are you still interested in finishing this PR?

**Comment by B-Galati** [comment]
Yes @fabpot 👍

**Comment by B-Galati** [comment]
@fabpot it's ready, the CI failure looks unrelated but I am may be wrong of course.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
@B-Galati Can you create a PR on symfony/recipes?

**Comment by fabpot** [comment]
Thank you @B-Galati.

**Comment by B-Galati** [comment]
PR for the recipe opened symfony/recipes#1113 :heavy_check_mark:

**Pull request merged**